api/rosetta/validator.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-playground/validator/v10"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+)
+
+var (
+	// Special case in the validator library that shouldn't really happen with correct usage.
+	errInvalidValidation = errors.New("invalid validation input")
+
+	errBlockLength   = errors.New(blockLength)
+	errAddressLength = errors.New(addressLength)
+	errTxLength      = errors.New(txLength)
+)
+
+// Field names needed by the validator library, not needed or used by our code at the moment.
+const (
+	blockHashField   = "block_hash"
+	blockchainField  = "blockchain"
+	networkField     = "network"
+	addressField     = "address"
+	txField          = "transaction_id"
+	currencyField    = "currency"
+	symbolField      = "symbol"
+	transactionField = "transaction"
+	signaturesField  = "signatures"
+)
+
+type Validator struct {
+	validate *validator.Validate
+}
+
+func NewValidator() *Validator {
+
+	v := validator.New()
+
+	// Register custom validators for known types.
+	v.RegisterStructValidation(blockValidator, identifier.Block{})
+	v.RegisterStructValidation(networkValidator, identifier.Network{})
+	v.RegisterStructValidation(accountValidator, identifier.Account{})
+	v.RegisterStructValidation(transactionValidator, identifier.Transaction{})
+	// Register custom top-level validators.
```

Comment:
These validators are for the whole request object, vs a specific type inside the required. This is so that you can validate standard types like strings, but still in a nice, structured way.

Reply:
As Brendan said, I agree these comments would add more value by being in the code.